File: minimumWindow.java
/* Time complexity: O(n)
space complexity: O(n)
*/


class Solution {
    public String minWindow(String s, String t) {
        HashMap<Character, Integer> map = new HashMap<>();
        for (char c : t.toCharArray())
            map.put(c, map.getOrDefault(c, 0) + 1);
        
        int totalFnd = t.length();
        String res = "";
        int len = Integer.MAX_VALUE;
        int j = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                int numChar = map.get(c) - 1;
                map.put(c, numChar);
                
                if (numChar >= 0)
                    totalFnd--;
            }
            
            while (totalFnd == 0) {
                if (i - j + 1 < len) {
                    res = s.substring(j, i + 1);
                    len = i - j + 1; 
                }

                char jChar = s.charAt(j);
                if (map.containsKey(jChar)) {
                    int numJChar = map.get(jChar) + 1;
                    map.put(jChar, numJChar);

                    if (numJChar > 0)
                        totalFnd++;
                }
                j++;
            }
        }
        
        return res;
    }
}

File: MinimumWindow.py
#All TC on leetocde passed


class Solution:
    def minWindow(self, s: str, t: str) -> str:

        #Here we use sliding window to find the t string in s
        #Time complexity - O(n) - traverse all chars of string s
        #Space complexity - O(1) - as max chars are only 52 alphabets
        t_map = collections.defaultdict(int)
        s_map = collections.defaultdict(int)
        

        for ch in t:
            t_map[ch]+=1

        matched = len(t_map)
        formed = 0

        res = ""
        resLen = 0

        l=0
        r=0

        while l<=r and r<len(s):

            ch = s[r]

            if ch in t_map:
                s_map[ch]+=1
            
            if ch in t_map and t_map[ch]==s_map[ch]:
                formed+=1
        
            
            while l<=r and formed==matched:

                ch = s[l]

                if res=="" or len(res)>(r-l+1):
                    res = s[l:r+1]
                
                if ch in t_map:
                    s_map[ch]-=1
                
                if ch in t_map and t_map[ch]>s_map[ch]:
                    formed-=1
                
                l+=1
            r+=1
            
        return res


        